src/v/storage/disk_log_impl.cc

```diff
@@ -507,40 +507,47 @@ segment_set disk_log_impl::find_sliding_range(
             // Stop once we get to an unstable segment.
             break;
         }
+        if (
+           _last_compaction_window_start_offset.has_value()
+           && (seg->offsets().get_base_offset()
+                >= _last_compaction_window_start_offset.value())) {
+            // Force clean segment production by compacting down to the
+            // start of the log before considering new segments in the
+            // compaction window.
+            break;
```

Comment:
It seems possible that segments could have been removed such that the first segment falls above _last_compaction_window_start_offset. In that case, does this always return empty and we get stuck never compacting?

Reply:
Maybe we should reset _last_compaction_window_start_offset at the top of this method. Then we wouldn't need to worry about resetting it at compaction time

Reply:
Good callout. Added a new sanity check to the top of `find_sliding_range()` to reset the `_last_compaction_window_start_offset` if it is `<= _segs.front()->offsets().get_base_offset()`.

[However, the code which resets it ](https://github.com/redpanda-data/redpanda/pull/23380/commits/4045b9f90f9be8a814c3bb15941f5b15bb690b64#diff-8caa0c9719de25bd209fa674959d9327cd55237de2fb4683761acf9bcc6de217R670) in `sliding_window_compact` is still required- we want to compare `idx_start_offset` to the base offset of the first segment in our sliding window range (which is not necessarily the front segment in the log) as an indicator of whether or not we can reset the start offset and allow new segments into the range.